raft.c

```diff
@@ -2343,27 +2345,73 @@ void handleShardGroupAdd(RedisRaftCtx *rr, RaftReq *req)
     RaftReqFree(req);
 }
 
+void handleShardGroupUpdate(RedisRaftCtx *rr, RaftReq *req)
+{
+    /* Must be done on a leader */
+    if (checkRaftState(rr, req) == RR_ERROR ||
+        checkLeader(rr, req, NULL) == RR_ERROR) {
+        RedisModule_ReplyWithError(req->ctx, "failed, please check logs.");
+        goto exit;
+    }
+
+    /* TODO: validate/handle slot range changes:
+     *  i.e. preventing stable overlap switching to migrating or adding as importing
+     */
+    ShardGroup *sg = getShardGroupById(rr, req->r.shardgroup_add.id);
+    if (sg == NULL) {
+        RedisModule_ReplyWithError(req->ctx, "failed, please check logs.");
+        goto exit;
+    }
+
+    if (compareShardGroups(sg, &req->r.shardgroup_add) != 0) {
+        if (ShardGroupAppendLogEntry(rr, &req->r.shardgroup_add,
+                                 RAFT_LOGTYPE_UPDATE_SHARDGROUP, req) == RR_ERROR) {
+            RedisModule_ReplyWithError(req->ctx, "failed, please check logs.");
+            goto exit;
+        }
+
+        /* wait till return till after update is applied */
+        return;
+    }
+
+    /* no changes, immediate return */
+    RedisModule_ReplyWithSimpleString(req->ctx, "OK");
+
+exit:
+    RaftReqFree(req);
+}
+
 /* Handles RAFT.SHARDGROUP GET which includes:
  * - Description of the local cluster as a shardgroup, including all nodes
  * - Description of remote shardgroups as last tracked.
  */
 
 void handleShardGroupGet(RedisRaftCtx *rr, RaftReq *req)
 {
-    int alen;
-
     /* Must be done on a leader */
     if (checkRaftState(rr, req) == RR_ERROR ||
         checkLeader(rr, req, NULL) == RR_ERROR) {
         goto exit;
     }
 
-    RedisModule_ReplyWithArray(req->ctx, REDISMODULE_POSTPONED_ARRAY_LEN);
-
-    RedisModule_ReplyWithLongLong(req->ctx, rr->config->sharding_start_hslot);
-    RedisModule_ReplyWithLongLong(req->ctx, rr->config->sharding_end_hslot);
-    alen = 2;
+    ShardGroup *sg = rr->sharding_info->shard_groups[0];
+    /* 2 arrays
+     * 1. slot ranges -> each element is a 3 element array start/end/type
+     * 2. nodes -> each element is a 2 element array id/address
+     */
+    RedisModule_ReplyWithArray(req->ctx, 3);
+    RedisModule_ReplyWithCString(req->ctx, redis_raft.snapshot_info.dbid);
+    RedisModule_ReplyWithArray(req->ctx, sg->slot_ranges_num);
+    for(int i = 0; i < sg->slot_ranges_num; i++) {
+        ShardGroupSlotRange * sr = &sg->slot_ranges[i];
+        RedisModule_ReplyWithArray(req->ctx, 3);
+        RedisModule_ReplyWithLongLong(req->ctx, sr->start_slot);
+        RedisModule_ReplyWithLongLong(req->ctx, sr->end_slot);
+        RedisModule_ReplyWithLongLong(req->ctx, sr->type);
```

Comment:
Consider either more explicit mapping of type to the wire format, or at least explicitly assign values to the enum - the way it's done now is a bit fragile IMHO.

Reply:
what do you mean by this?  that I should validate the values?  what would you recommend?

Reply:
I mean use specific values in the enum and have a comment that indicate this is not just an in-memory enum, but something that gets serialized and therefore must be stable (i.e. nobody should reorder elements).

Reply:
so basically a switch there over all the values of the enum as a form of validation? (or just have fallthrough for all valid enum values and a panic in default?

Reply:
I think at least bound check should be done.

Reply:
something like this?

```
        assert(sr->type == SLOTRANGE_STABLE ||
              sr->type == SLOTRANGE_MIGRATING ||
              sr->type == SLOTRANGE_IMPORTING);
```